### ISSUE TITLE
Tensor identity comparisons not working / unclear

### ISSUE THREAD

**Issue by FlorianDietz** [pytorch/pytorch#20947]
## 🐛 Bug

Performing identity comparisons between tensors does not work as expected. It looks like 'is' is overloaded but I can't find documentation to that effect.

## To Reproduce

I want to compare two tensors by identity / by reference. Normally, comparing two objects in python by identity works with ‘is’ or by comparing id() values. However, this does not seem to work:

a = torch.tensor([1,2,3])
b = a.detach()
print(a is b) # False (good, since b is derived from a but not identical to it)
print(a.data is b.data) # False (bad, since they refer to exactly the same thing)
print(id(a.data) == id(b.data)) # True (good, since they refer to exactly the same thing)

Shouldn’t either both of the last two statements be True, or neither? It seems odd that 'is' gives a different result than id()==id().

It looks like I can make a reference comparison with id()==id(), but I am reluctant to trust that this will always work because to my knowledge using ‘is’ should be the way to do it, and that isn’t working.

Can anyone clarify if that is a bug or a feature? How are you _supposed_ to compare two tensors by reference?

**Comment by colesbury** [comment]
Tensor.data returns an alias, just like detach() -- same underlying data but different Tensor object. Two evaluations of `x.data` do **not** refer to the same thing.

> Shouldn’t either both of the last two statements be True, or neither? It seems odd that 'is' gives a different result than id()==id().

No, this is because `id` returns the memory address and memory addresses can be reused once an object goes out scope. `a.data` goes out of scope before `b.data` is evaluated. `id` only uniquely identify an object **during that object's lifetime!** The same often happens with core Python objects:

```
id({}) == id({})  # True
{} is {}  # False
```

If you keep the object in scope the id is unique:

```
tmp0 = a.data
tmp1 = b.data
id(tmp0) == id(tmp1)  # False
```

**Comment by colesbury** [comment]
> How are you supposed to compare two tensors by reference?

Use `is` (like other Python objects)

**Comment by FlorianDietz** [comment]
Thanks for the explanation about id(), it makes sense to me now.

Coming back to the example though: Is there any way to compare tensors so that `a` and `b` from the example are recognized as the same thing? They refer to the same data, but one of them is detached while the other is the original.

**Comment by colesbury** [comment]
> Is there any way to compare tensors so that a and b from the example are recognized as the same thing?

You can check that two Tensor's data_ptr(), shape, and stride() match, but this sound pretty fragile and you should probably instead rethink your use-case.
